x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatePeriod.java

```diff
@@ -0,0 +1,156 @@
+    ) {
+        super(source, v);
+    }
+
+    private ToDatePeriod(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected Map<DataType, BuildFactory> factories() {
+        return EVALUATORS;
+    }
+
+    @Override
+    protected final TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        TypeResolution typeResolution = isType(
+            field(),
+            factories()::containsKey,
+            sourceText(),
+            FIRST,
+            supportedTypesNames(supportedTypes())
+        );
+        if (field().foldable()) {
+            if (isString(field().dataType()) && isValidInterval(field().fold().toString()) == false) {
+                typeResolution = new TypeResolution(LoggerMessageFormat.format(null, INVALID_INTERVAL_ERROR, sourceText(), field().fold()));
+            }
+
+        }
+        return typeResolution;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DATE_PERIOD;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new ToDatePeriod(source(), newChildren.get(0));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, ToDatePeriod::new, field());
+    }
+
+    @Override
+    public final Object fold() {
+        if (field() instanceof Literal l) {
+            Object v = l.value();
+            if (v instanceof BytesRef b) {
+                if (isValidInterval(b.utf8ToString()) == false) {
+                    throw new IllegalArgumentException(
+                        LoggerMessageFormat.format(null, INVALID_INTERVAL_ERROR, sourceText(), b.utf8ToString())
+                    );
+                }
+                return EsqlDataTypeConverter.parseTemporalAmount(b.utf8ToString(), DATE_PERIOD);
+            }
+        } else if (field() instanceof ToDatePeriod && field().foldable()) {
+            return field().fold();
+        }
+
+        throw new IllegalArgumentException(
+            LoggerMessageFormat.format(
+                null,
+                "{}argument of [{}] must be a constant, received [{}]",
+                FIRST.name().toLowerCase(Locale.ROOT) + " ",
+                sourceText(),
+                Expressions.name(field())
+            )
+        );
+    }
+
+    private boolean isValidInterval(String interval) {
+        String[] input = interval.toLowerCase(Locale.ROOT).stripLeading().stripTrailing().split("\\s+");
+        if (input.length != 2 || DATE_PERIODS.contains(input[1]) == false) {
+            return false;
+        }
+        try {
+            Integer.parseInt(input[0].toString());
+        } catch (NumberFormatException e) {
+            return false;
+        }
+        return true;
+    }
```

Comment:
This duplicates the parsing of date periods from `EsqlDataTypeConverter`. I think there should be only 1 place that defines how date periods are parsed.

Suggestion: in `fold` we can try calling `parseTemporalAmount` and catch the parsing exception; in `resolveType` we can just not validate the folded constant - the query will fail at constant folding time - this should be consistent with `Percentile`.

Applies similarly to `ToTimeDuration`.

Reply:
Yeah, there are some redundancy here, I slightly leaning toward validate the format as soon as possible, let me see if I can refactor it a bit to reduce the redundancy here.